aws/rust-runtime/aws-config/src/sts/assume_role.rs

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Assume credentials for a role through the AWS Security Token Service (STS).
+
+use aws_sdk_sts::error::AssumeRoleErrorKind;
+use aws_sdk_sts::{Client, Config};
+use aws_types::credentials::{self, future, CredentialsError, ProvideCredentials};
+use aws_types::region::Region;
+
+use tracing::Instrument;
+
+/// Credentials provider that uses credentials provided by another provider to assume a role
+/// through the AWS Security Token Service (STS).
+///
+/// When asked to provide credentials, this provider will first invoke the inner credentials
+/// provider to get AWS credentials for STS. Then, it will call STS to get assumed credentials for
+/// the desired role.
+///
+/// # Examples
+/// ```rust
+/// use aws_config::sts::{AssumeRoleProvider};
+/// use aws_types::{Credentials, region::Region};
+/// use aws_config::environment;
+/// use aws_config::environment::credentials::EnvironmentVariableCredentialsProvider;
+///
+/// let provider = AssumeRoleProvider::new(
+///     EnvironmentVariableCredentialsProvider::new(),
+///     "arn:aws:iam::123456789012:role/demo",
+///     Region::from_static("us-east-2"),
+///     "testAR",
+/// );
+/// ```
+#[derive(Debug, Clone)]
+pub struct AssumeRoleProvider<P> {
+    provider: P,
+    role: String,
+    session: String,
+    region: Region,
+}
+
+impl<P> AssumeRoleProvider<P>
+where
+    P: ProvideCredentials + Clone + 'static,
+{
+    /// Use the given `provider` to assume the given `role`.
+    ///
+    /// `role` is the Amazon Resource Name (ARN) of the role to assume.
+    /// `session` is a temporary identifier for the assumed role session.
+    pub fn new(
+        provider: P,
+        role: impl Into<String>,
+        region: Region,
+        session: impl Into<String>,
+    ) -> Self {
+        Self {
+            provider,
+            role: role.into(),
+            region,
+            session: session.into(),
+        }
+    }
+
+    async fn credentials(&self) -> credentials::Result {
+        let span = tracing::info_span!("assume_role",
+            arn = %self.role,
+            session = %self.session,
+        );
+        // NOTE: Do not .await before dropping this guard.
+        let span_guard = span.enter();
+
+        tracing::info!("assuming role");
+
+        let config = Config::builder()
+            .credentials_provider(self.provider.clone())
+            .region(self.region.clone())
+            .build();
+        let client = Client::from_conf(config);
```

Comment:
this will actually recreate the connection every time (which is definitely not what we want, especially since constructing a native-tls Rust connection actually takes 300ms on OSX)

You need to keep the inner client around, then use the low-level API to late-bind the config.

Reply:
Yeah, good point. Do you have a reference for late-binding the config?

Reply:
https://github.com/awslabs/smithy-rs/blob/0b32a897ed0ac9252b2bb2485b1258ac0cfd0760/aws/rust-runtime/aws-config/src/profile/credentials/exec.rs#L42-L65

Reply:
Done. Though I went a step further and stored the entire client in `AssumeRoleProvider`. That _did_ mean adding a method `low_level_client` to get at the `smithy_client::Client` from an SDK client though, which I don't know whether or not you're okay with? I ideally wanted to go a step further and include the entire generated operation (so the result of `make_operation`) in there too, but didn't know if that was actually sound to do?

Reply:
I removed `low_level_client` again and stored a `StandardClient` instead.